### DISCUSSION TITLE
'Magick' item empty if Magick::InitializeMagick called

### DISCUSSION THREAD

**Discussion by adhawkins** [ImageMagick/ImageMagick#5864] [Help]
If I run the following code against the attached image, I get strange results. If I *don't* call 'InitializeMagick', I get an appropriate value back for the 'magick()' string. However, if I do call it, I get an empty string:

```
#include <iostream>

#include <Magick++.h>

#include <fstream>
#include <vector>

int main(int argc, const char *argv[])
{
	if (argc == 2 || argc == 3)
	{
		if (argc == 3)
		{
			std::cout << "Initialising" << std::endl;
			Magick::InitializeMagick(NULL);
		}

		std::ifstream Stream(argv[1], std::ios::in | std::ios::binary);
		std::vector<uint8_t> Contents((std::istreambuf_iterator<char>(Stream)), std::istreambuf_iterator<char>());

		Magick::Blob Blob(&Contents[0], Contents.size());
		Magick::Image Image(Blob);

		std::cout << "Magick: '" << Image.magick() << "'" << std::endl;
	}
	else
	{
		std::cout << "Usage: " << argv[0] << " filename [initialise]" << std::endl;
	}
}
```
```
[andy@xcp-dev$ cmake --build build && build/imagetest nirvana.png
Magick: 'PNG'
[andy@xcp-dev$ cmake --build build && build/imagetest nirvana.png a
Initialising
Magick: ''
```

I'm running under Debian stable in an AMD64 VM.

Can anyone offer any suggestions as to why this might be?

Andy
![nirvana](https://user-images.githubusercontent.com/1628607/206401764-e0aaf076-7870-4e7a-9542-08773714c756.png)

**Comment by urban-warrior** [comment]
We're using ImageMagick 7.1.0-53 and your codeset returns `PNG` for both instances of your command-line, as expected.

**Comment by adhawkins** [comment]
Debian reports the version as '6.9.11.60+dfsg-1.3'. They don't seem to have version 7.

Anyone able to test it using Debian to make sure it's not me doing something stupid?

I'm building using cmake, this is the command line it's generating:

```
[1/2] /usr/bin/g++ -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -I. -I../ -isystem /usr/include/ImageMagick-6 -isystem /usr/include/x86_64-linux-gnu/ImageMagick-6 -g -Wall -Wextra -pedantic -Werror -std=gnu++11 -MD -MT CMakeFiles/imagetest.dir/imagetest.cc.o -MF CMakeFiles/imagetest.dir/imagetest.cc.o.d -o CMakeFiles/imagetest.dir/imagetest.cc.o -c ../imagetest.cc
[2/2] : && /usr/bin/g++ -g  CMakeFiles/imagetest.dir/imagetest.cc.o -o imagetest  /usr/lib/x86_64-linux-gnu/libMagick++-6.Q16.so && :
```

**Comment by snibgo** [comment]
@adhawkins: Have you tried `InitializeMagick(*argv);` ?

**Comment by adhawkins** [comment]
I thought that was only for Windows platforms?

Just tried it, same result.

**Comment by adhawkins** [comment]
Just another thought @urban-warrior , could you try again, downloading the image from here:

http://coverartarchive.org/release/653988c5-c79f-4b24-840a-6b1d21144405/10224339003.png

Just in case GitHub have reformatted it or something during the upload process? That URL is the source of the image I'm working with.

**Comment by urban-warrior** [comment]
We tried your program again, this time with ImageMagick 6.9.12-68, and again it worked as expected with your sample image.  It returns `Magick: 'PNG'` with and without initializing.  It should not matter, but this is our build command:

```
c++ `Magick++-config --cxxflags --cppflags` -O2 -o demo demo.cpp `Magick++-config --ldflags --libs`
```

**Reply by adhawkins** [reply]
@urban-warrior What does 'Magick++-config --cxxflags --cppflags' return for you? I wonder if it's the values for MAGICKCORE_HDRI_ENABLE or MAGICKCORE_QUANTUM_DEPTH that are causing the issue?

**Comment by adhawkins** [comment]
Ok, I spun up an Alpine container and built on there. It has ImageMagick 7.1.0.52. The program works as expected, whether or not I initialize.

I guess I should put some logic around the call initialize to exclude it if the version is below a certain number. Any suggestions on how I can work out what that number is?

**Reply by urban-warrior** [reply]
See `MagickCore/version.h`, specifically `MagickLibVersion`.

**Reply by adhawkins** [reply]
Yes, but I wonder how I can work out which version 'works'? It appears that the version in Debian fails if I call initialise, but the one you tried (which is just later than the Debian one) doesn't. Is there likely to be a fix for this that I can find to see when it was included?

**Reply by urban-warrior** [reply]
We don't have a clue if the issue is transient or if there was a particular patch in the past that fixed the issues.  We often use binary search across all versions to trace a defect.  You can find older source releases @ https://imagemagick.org/archive/releases/.

**Reply by adhawkins** [reply]
Ok, understood. Will have to investigate what's involved in building ImageMagick myself, and see if I can find whether this is an issue specific to Debian or, if not, which version(s) it affected.

Thanks for the assistance, appreciate it.

**Comment by adhawkins** [comment]
Ok, I built the same version that Debian claims to be providing in an Alpine container, and both samples work fine. I'll see if it's reproducible on a 'clean' Debian install and report it as a bug to them if I can.

Thanks everyone for the assistance.
